File: TS/reshape.ts
// Retourne la chaine formatee en lignes de n caracteres maximum (mots non coupes)
function reshape(n: number, str: string): string {
  if (n <= 0) {
    throw new Error("n must be greater than 0");
  }

  const words = str.split(" ");
  const lines: string[] = [];
  let line = "";

  for (const word of words) {
    if (line.length + word.length + 1 > n) {
      lines.push(line.trim());
      line = "";
    }
    line += word + " ";
  }
  if (line.length > 0) {
    lines.push(line.trim());
  }

  return lines.join("\n");
}

export { reshape };

File: Java/src/Reshape.java
public class Reshape {

    static void main() {
        String str = "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.";
        int n = 20;
        String reshapedStr = reshape(n, str);
        System.out.println(reshapedStr);
    }

    /***
     * la methode reshape doit retourner la chaîne str sans les espaces et formattée en lignes
     * de n caractères maximum
     * @param n
     * @param str
     * @return
     */
    public static String reshape(int n, String str) {
        if (n <= 0) {
            throw new IllegalArgumentException("n must be greater than 0");
        }
        StringBuilder result = new StringBuilder();
        String[] words = str.split(" ");
        StringBuilder line = new StringBuilder();

        for (String word : words) {
            if (line.length() + word.length() + 1 > n) {
                result.append(line.toString().trim()).append("\n");
                line.setLength(0);
            }
            line.append(word).append(" ");
        }
        if (line.length() > 0) {
            result.append(line.toString().trim());
        }

        return result.toString();
    }
}
